File: varieties/extract-variants.pl
use strict;
use warnings;
use autodie;
use utf8;

binmode( STDOUT, ":utf8" );

my @files = ( "adjectives-fdic.txt", "nouns-fdic.txt", "verbs-fdic.txt");
my %lemmas;
my $file;

foreach $file (@files) {
	open(my $fh,  "<:encoding(UTF-8)", "../src-dict/".$file);
		while (my $line = <$fh>) {
	    chomp($line);
	    if ($line =~ /^([^# =]+)/) {       
	        my $lemma=$1;
	        $lemmas{$lemma}=1;
	    }
	}
	close ($fh);
}

@files = ( "adverbs-lt.txt", "adv_mente-lt.txt", "propernouns-lt.txt", "resta-lt.txt");
foreach $file (@files) {
	open(my $fh,  "<:encoding(UTF-8)", "../src-dict/".$file);
		while (my $line = <$fh>) {
	    chomp($line);
	    if ($line =~ /^.* (.*) .*$/) {       
	        my $lemma=$1;
	        $lemmas{$lemma}=1;
	    }
	}
	close ($fh);
}

open(my $ofh,  ">:encoding(UTF-8)", "br-pt_excluding.txt");
open(my $ofh2,  ">:encoding(UTF-8)", "br-pt_recommend-BR.txt");
open(my $ofh3,  ">:encoding(UTF-8)", "br-pt_uncertain.txt");

foreach my $lemma (sort keys %lemmas) {
	if ($lemma =~ /ê/) {
		my $lemma2 = $lemma;
		$lemma2 =~ s/ê/é/;
		if (exists($lemmas{$lemma2})) {
			print $ofh "$lemma=$lemma2\n";
		}
	}
	if ($lemma =~ /ô/) {
		my $lemma2 = $lemma;
		$lemma2 =~ s/ô/ó/;
		if (exists($lemmas{$lemma2})) {
			print $ofh "$lemma=$lemma2\n";
		}
	}

	if ($lemma =~ /pt/) {
		my $lemma2 = $lemma;
		$lemma2 =~ s/pt/t/;
		if (exists($lemmas{$lemma2})) {
			print $ofh2 "$lemma2=$lemma\n";
		}
	}
	if ($lemma =~ /cç/) {
		my $lemma2 = $lemma;
		$lemma2 =~ s/cç/ç/;
		if (exists($lemmas{$lemma2})) {
			print $ofh2 "$lemma2=$lemma\n";
		}
	}

	if ($lemma =~ /ct/) {
		my $lemma2 = $lemma;
		$lemma2 =~ s/ct/t/;
		if (exists($lemmas{$lemma2})) {
			print $ofh3 "$lemma2=$lemma\n";
		}
	}
	# recepção
	if ($lemma =~ /pç/) {
		my $lemma2 = $lemma;
		$lemma2 =~ s/pç/ç/;
		if (exists($lemmas{$lemma2})) {
			print $ofh3 "$lemma=$lemma2\n";
		}
	}
}

close($ofh);

File: variants/extract_variants.py
"""Extract differences between PT varieties."""
from re import Pattern

from tqdm import tqdm
import re
from typing import List

from variants.syllabifier import Syllabifier
from variants.alternation import AlternationContext, Alternation

SYLLABIFIER = Syllabifier()

FILES_TO_PROCESS_COMPLEX = {
    "adjectives-fdic.txt", "nouns-fdic.txt", "verbs-fdic.txt"
}

FILES_TO_PROCESS_SIMPLE = {
    "adverbs-lt.txt", "adv_mente-lt.txt", "propernouns-lt.txt", "resta-lt.txt"
}

alternations = [
    Alternation(AlternationContext('ê', 'vowel', SYLLABIFIER), 'é', []),
    Alternation(AlternationContext('ô', 'vowel', SYLLABIFIER), 'ó', []),
    Alternation(AlternationContext('pt', 'contains', SYLLABIFIER), 't', []),
    Alternation(AlternationContext('cç', 'contains', SYLLABIFIER), 'ç', []),
    Alternation(AlternationContext('ct', 'contains', SYLLABIFIER), 't', []),
    Alternation(AlternationContext('pç', 'contains', SYLLABIFIER), 'ç', []),
]


def process_file(filename: str, pattern: Pattern) -> List[str]:
    """Define a function to process a file and collect lemmata."""
    lemmata = []
    with open(f"../src-dict/{filename}", 'r', encoding='utf-8') as file:
        for line in file:
            line = line.strip()
            match = re.match(pattern, line)
            if match:
                lemmata.append(match.group(1))
    return lemmata


def collect_lemmata() -> set[str]:
    lemmata = []
    for filepath in FILES_TO_PROCESS_SIMPLE:
        lemmata.extend(process_file(filepath, re.compile('^.* (.*) .*$')))
    for filepath in FILES_TO_PROCESS_COMPLEX:
        lemmata.extend(process_file(filepath, re.compile('^([^# =]+)')))
    return set(lemmata)


LEMMATA = sorted(collect_lemmata())
for lemma in tqdm(LEMMATA):
    for alternation in alternations:
        try:
            alternation.transform(lemma)
        except TypeError as e:
            print(f"\"{lemma}\": {e}")

TRANSFORMATIONS = []
for alternation in alternations:
    transformations = filter(lambda transformation: transformation.target in LEMMATA, alternation.transformations)
    TRANSFORMATIONS.extend(transformations)

with open('br-pt.txt', 'a', encoding='utf-8') as output_file:
    output_file.write("\n".join([transformation.__str__() for transformation in TRANSFORMATIONS]))
